monoio/src/blocking.rs

```diff
@@ -0,0 +1,273 @@
+//! Blocking tasks related.
+
+use std::{cell::RefCell, future::Future, sync::Arc, task::Poll};
+
+use fxhash::FxHashMap;
+
+use crate::{
+    runtime::Context,
+    task::{new_task, JoinHandle},
+    utils::thread_id::BLOCKING_THREAD_ID,
+};
+
+/// Users may implement a ThreadPool and attach it to runtime.
+pub trait ThreadPool {
+    /// Monoio runtime will call `schedule_task` on `spawn_blocking`.
+    /// ThreadPool impl must execute it now or later.
+    fn schedule_task(&self, task: BlockingTask);
+}
+
+/// Error on waiting blocking task.
+#[derive(Debug, Clone, Copy)]
+pub enum JoinError {
+    /// Task is canceled.
+    Canceled,
+}
+
+/// BlockingTask is contrusted by monoio, ThreadPool impl
+/// will exeucte it with `.run()`.
+pub struct BlockingTask {
+    task: Option<crate::task::Task<NoopScheduler>>,
+    blocking_vtable: &'static BlockingTaskVtable,
+}
+
+unsafe impl Send for BlockingTask {}
+
+struct BlockingTaskVtable {
+    pub(crate) drop: unsafe fn(&mut crate::task::Task<NoopScheduler>),
+}
+
+fn blocking_vtable<V>() -> &'static BlockingTaskVtable {
+    &BlockingTaskVtable {
+        drop: blocking_task_drop::<V>,
+    }
+}
+
+fn blocking_task_drop<V>(task: &mut crate::task::Task<NoopScheduler>) {
+    let mut opt: Option<Result<V, JoinError>> = Some(Err(JoinError::Canceled));
+    unsafe { task.finish((&mut opt) as *mut _ as *mut ()) };
+}
+
+impl Drop for BlockingTask {
+    fn drop(&mut self) {
+        if let Some(task) = self.task.as_mut() {
+            unsafe { (self.blocking_vtable.drop)(task) };
+        }
+    }
+}
+
+// scoped_thread_local!(static BLOCKING_CTX: Context);
+thread_local! {
+    static BLOCKING_CTX: Context = Context {
+        thread_id: BLOCKING_THREAD_ID,
+        unpark_cache: RefCell::new(FxHashMap::default()),
+        waker_sender_cache: RefCell::new(FxHashMap::default()),
+        tasks: Default::default(),
+        time_handle: None,
+        blocking_handle: BlockingHandle::Empty(BlockingStrategy::Panic),
+    };
+}
+
+impl BlockingTask {
+    /// Run task.
+    #[inline]
+    pub fn run(mut self) {
+        let task = self.task.take().unwrap();
+        // if we are within a runtime, just run it.
+        if crate::runtime::CURRENT.is_set() {
```

Comment:
It will use current thread to execute it. This behavior maybe bad so it's not by default. This is designed only for users that absolutely do not want to create a thread pool.

Reply:
I think we need try to avoid implicit behavior, it's difficult to find bug